source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -244,14 +271,112 @@ bool Filter::parseExtensions(os_fd_t fd) {
     if (bytes_avail == 0) {
       return false;
     }
-    bytes_avail = std::min(size_t(bytes_avail), sizeof(buf_));
+    bytes_avail = std::min(size_t(bytes_avail), buf_size);
     bytes_avail = std::min(size_t(bytes_avail), proxy_protocol_header_.value().extensions_length_);
-    const Api::SysCallSizeResult recv_result = os_syscalls.recv(fd, buf_, bytes_avail, 0);
+    buf += buf_off ? *buf_off : 0;
+    const Api::SysCallSizeResult recv_result = os_syscalls.recv(fd, buf, bytes_avail, 0);
     if (recv_result.rc_ != bytes_avail) {
       throw EnvoyException("failed to read proxy protocol extension");
     }
     proxy_protocol_header_.value().extensions_length_ -= recv_result.rc_;
+
+    if (buf_off) {
+      *buf_off += recv_result.rc_;
+    }
+  }
+
+  return true;
+}
+
+/**
+ * @note  A TLV is arranged in the following format: @n
+ *        struct pp2_tlv {      @n
+            uint8_t type;       @n
+            uint8_t length_hi;  @n
+            uint8_t length_lo;  @n
+            uint8_t value[0];   @n
+          };                    @n
+ *        See https://www.haproxy.org/download/2.1/doc/proxy-protocol.txt for details
+ */
+void Filter::parseTlvs(const std::vector<uint8_t>& tlvs) {
+  size_t size = tlvs.size();
+  uint8_t tlv_type;
+  uint8_t tlv_length_upper;
+  uint8_t tlv_length_lower;
+  size_t tlv_value_length;
+  size_t idx{0};
+  while (idx < size) {
+    // get type
+    tlv_type = tlvs[idx];
+    idx++;
+
+    // get length
+    if ((idx + 1) >= size) {
+      throw EnvoyException(
+          fmt::format("failed to read proxy protocol extension. No bytes for TLV length. "
+                      "Extension length is {}, current index is {}, current type is {}.",
+                      size, idx, tlv_type));
+    }
+
+    tlv_length_upper = tlvs[idx];
+    tlv_length_lower = tlvs[idx + 1];
+    tlv_value_length = (tlv_length_upper << 8) + tlv_length_lower;
+    idx += 2;
+
+    // get the value
+    if ((idx + tlv_value_length - 1) >= size) {
+      throw EnvoyException(
+          fmt::format("failed to read proxy protocol extension. No bytes for TLV value. "
+                      "Extension length is {}, current index is {}, current type is {}, current "
+                      "value length is {}.",
+                      size, idx, tlv_type, tlv_length_upper));
+    }
+
+    // only save to dynamic metadata if this type of TLV is needed
+    std::string filter_name = ListenerFilterNames::get().ProxyProtocol;
+    auto key_value_pair = config_->isTlvTypeNeeded(tlv_type);
+    if (nullptr != key_value_pair) {
+      std::string tlv_value(reinterpret_cast<char const*>(tlvs.data() + idx), tlv_value_length);
+      auto metadata_value = ProtobufWkt::Value();
+      metadata_value.set_string_value(std::move(tlv_value));
+
+      std::string metadata_key = key_value_pair->metadata_namespace().empty()
+                                     ? filter_name
```

Comment:
replace with `ListenerFilterNames::get().ProxyProtocol` and remove variable `filter_name` to avoid an allocation of the string.

Reply:
updated.